File: src/d1hashmaps/JS/DesignHashMap.js
const Bucket = require('./Bucket');

class DesignHashMap {
    constructor() {
        this.keySpace = 2069;
        this.buckets = Array.from({ length: this.keySpace }, () => new Bucket());
    }

    put(key, value) {
        const hashKey = key % this.keySpace;
        this.buckets[hashKey].update(key, value);
    }

    get(key) {
        const hashKey = key % this.keySpace;
        return this.buckets[hashKey].get(key);
    }

    remove(key) {
        const hashKey = key % this.keySpace;
        this.buckets[hashKey].remove(key);
    }

    printHashMap() {
        this.buckets.forEach((bucket, i) => {
            if (bucket.pairs.length > 0) {
                const pairs = bucket.pairs.map(pair => `{key: ${pair.key}, value: ${pair.value}}`).join(' ');
                console.log(`Bucket ${i}: ${pairs}`);
            }
        });
    }
}

module.exports = DesignHashMap;

File: src/d1hashmaps/Python/DesignHashMap.py
from Bucket import Bucket


class DesignHashMap:
    def __init__(self):
        self.key_space = 2069
        self.buckets = [Bucket() for _ in range(self.key_space)]

    def put(self, key, value):
        hash_key = key % self.key_space
        self.buckets[hash_key].update(key, value)

    def get(self, key):
        hash_key = key % self.key_space
        return self.buckets[hash_key].get(key)

    def remove(self, key):
        hash_key = key % self.key_space
        self.buckets[hash_key].remove(key)

    def print_hash_map(self):
        for i, bucket in enumerate(self.buckets):
            if bucket.pairs:
                pairs_str = ', '.join(
                    f'{{key: {pair.key}, value: {pair.value}}}' for pair in bucket.pairs)
                print(f'Bucket {i}: {pairs_str}')
